DSCResources/MSFT_xADUser/MSFT_xADUser.psm1

```diff
@@ -1,229 +1,964 @@
+function Get-ADCommonParameters
+{
+    [CmdletBinding()]
+    [OutputType([System.Collections.Hashtable])]
+    param
+    (
+        [Parameter(Mandatory = $true)]
+        [ValidateNotNullOrEmpty()]
+        [System.String]
+        $UserName,
+
+        [ValidateNotNullOrEmpty()]
+        [System.String]
+        $CommonName,
+
+        [ValidateNotNullOrEmpty()]
+        [System.Management.Automation.PSCredential]
+        $DomainAdministratorCredential,
+
+        [ValidateNotNullOrEmpty()]
+        [System.String]
+        $DomainController,
+
+        [Parameter(ValueFromRemainingArguments)]
+        $IgnoredArguments,
+
+        [System.Management.Automation.SwitchParameter]
+        $UseNameParameter
+    )
+    
+    ## The Get-ADUser, Set-ADUser and Remove-ADUser cmdlets take an -Identity parameter, but the New-ADUser cmdlet uses the -Name parameter
+    if ($UseNameParameter)
+    {
+        if ($PSBoundParameters.ContainsKey('CommonName'))
         {
-            if ($Ensure -ne "Absent")
-            {
-                $params = @{
-                    Name = $UserName
-                    Credential = $DomainAdministratorCredential
-                    Enabled = $true
-                    UserPrincipalName = "$UserName@$DomainName"
-                    PasswordNeverExpires = $true
-                }
-                if ($Password)
-                {
-                    $params.Add( "AccountPassword", $Password.Password )
-                }
-                New-AdUser @params
-                Write-Verbose -Message "Successfully created user account '$($UserName)' in domain '$($DomainName)'."
-            }
+            $adUserParameters = @{ Name = $CommonName; }
         }
         else
         {
-            return ($Ensure -eq "Absent")
+            $adUserParameters = @{ Name = $UserName; }
         }
     }
+    else
+    {
+        $adUserParameters = @{ Identity = $UserName; }
+    }
+
+    if ($DomainAdministratorCredential)
+    {
+        $adUserParameters['Credential'] = $DomainAdministratorCredential;
+    }
+    if ($DomainController)
+    {
+        $adUserParameters['Server'] = $DomainController;
+    }
+    return $adUserParameters;
+
+} #end function Get-ADCommonParameters
+
+# Internal function to assert if the role specific module is installed or not
+function Assert-Module
+{
+    [CmdletBinding()]
+    param
+    (
+        [System.String] $ModuleName = 'ActiveDirectory'
+    )
+
+    if (-not (Get-Module -Name $ModuleName -ListAvailable))
+    {
+        $errorId = 'xADUser_ModuleNotFound';
+        $errorMessage = $LocalizedData.RoleNotFoundError -f $moduleName;
+        ThrowInvalidOperationError -ErrorId $errorId -ErrorMessage $errorMessage;
+    }
+} #end function Assert-Module
+
+# Internal function to get an Active Directory object's parent Distinguished Name
+function Get-ADObjectParentDN
+{
+    # https://www.uvm.edu/~gcd/2012/07/listing-parent-of-ad-object-in-powershell/
```

Comment:
hi @iainbrighton,
Unfortunately the blog post above does not contain information about the license it uses.
I've followed up with our legal team and even though you only used few lines of code from there, you would need to get permission of the blog author to include it into your contribution. Could you please follow up with the author to get such permission?
Thanks!